sim/src/main/cc/bridges/UARTModule.cc

```diff
@@ -0,0 +1,84 @@
+// See LICENSE for license details.
+
+#ifndef RTLSIM
+#include "simif_f1.h"
+#define SIMIF simif_f1_t
+#else
+#include "simif_emul.h"
+#define SIMIF simif_emul_t
+#endif
+
+#include "bridges/uart.h"
+#include "simif_peek_poke.h"
+
+class UARTModuleTest final : public simif_peek_poke_t {
+public:
+  class Handler final : public uart_handler {
+  public:
+    Handler(UARTModuleTest &test) { test.handler = this; }
+
+    ~Handler() {
+      // Check that the input and output buffers are equal.
+      if (in_buffer != out_buffer) {
+        fprintf(stderr,
+                "Buffer mismatch:\n  %s\n  %s\n",
+                in_buffer.c_str(),
+                out_buffer.c_str());
+        abort();
+      }
+    }
+
+    std::optional<char> get() override {
+      if (in_index >= in_buffer.size())
+        return std::nullopt;
+      return in_buffer[in_index++];
+    }
+
+    void put(char data) override { out_buffer += data; }
+
+    const std::string in_buffer = "We are testing the UART bridge";
+    std::string out_buffer;
+    size_t in_index = 0;
+  };
+
+  Handler *handler;
+
+  UARTModuleTest(simif_t &simif)
+      : simif_peek_poke_t(&simif, PEEKPOKEBRIDGEMODULE_0_substruct_create),
+        uart(std::make_unique<uart_t>(&simif,
+                                      UARTBRIDGEMODULE_0_substruct_create,
+                                      std::make_unique<Handler>(*this))) {}
+
+  void run() {
+    // Initialise the UART bridge.
+    uart->init();
+
+    // Reset the device.
+    poke(reset, 1);
+    step(1);
+    poke(reset, 0);
+    step(1);
+
+    // Tick until the out buffer is filled in.
+    step(300000, false);
+    while (!simif->done()) {
```

Comment:
Better, now you should add an iteration bound in case something deadlocks in the target. YOu could instead check that the DUT is making forward progress by calling simif_t::actual_tcycle, but that's a little trickier to tune -- i'll leave it to your discretion. 

This sort of thing can be caught by a host-time max cycle bound in the metasim harness, but you can provide a better error here.